Adding write-up for monitoring using Open Telemetry (#867)
Azure-Samples/modern-data-warehouse-dataops: docs/Observability_using_OpenTelemetry.md

# Observability in Data Projects Using OpenTelemetry

> 

Reviewer: Remove the following TODOs to merge the PR:

1. the 1st and 2nd of updating sample code link will be resolved in this PR. #933 
2. the others will **NOT** be included in this PR but be tracked with issues #937 , #936 , #935 , #934

> TO DO:

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

> - Update notebook links which are good examples for OTEL implementation.

Reviewer: ```suggestion
```

> - Replace the sample contents for python scripts (otel_monitor_invoker.py and otel_collector_invoker.py) with links.

Reviewer: ```suggestion
```

> - Where do we keep the config file related to OTEL

Reviewer: ```suggestion
```

> - What is the preferred setup for Collector setup (VM + Docker)?

Reviewer: ```suggestion
```

> - Including the monitoring module as part of the Fabric envs

Reviewer: ```suggestion
```

> - Code sample section updates - Need to add authentication and extensions - Check [Related feature](https://github.com/Azure-Samples/modern-data-warehouse-dataops/issues/715)

Reviewer: ```suggestion
```

Furthermore, this article delves into the implementation of monitoring and observability through the utilization of an open-source and vendor-agnostic framework known as [OpenTelemetry [OTEL]](https://opentelemetry.io/docs/what-is-opentelemetry/). There are various facets of monitoring, some of which are highlighted below, along with the corresponding services provided by Microsoft Azure:

- Application monitoring
  - Performance
  - Business metrics/Data quality
- Service monitoring
  - [Azure Monitor](https://learn.microsoft.com/azure/azure-monitor/)
  - [Microsoft Fabric metrics](https://learn.microsoft.com/fabric/enterprise/metrics-app)
  - [Microsoft Fabric monitoring workspace](https://learn.microsoft.com/fabric/admin/monitoring-workspace)
  - [Microsoft Fabric monitoring hub](https://learn.microsoft.com/fabric/admin/monitoring-hub)
- Security monitoring
  - [User activity tracking](https://learn.microsoft.com/fabric/admin/track-user-activities)

This article focuses on application monitoring (user-generated telemetry).

## Monitoring and observability in data projects

> Customers heavily rely on dashboards and metrics to evaluate the **business value** offered by a service or process. Therefore, it is crucial to capture all the necessary information at the appropriate granularity and frequency in order to generate these customer-facing dashboards. These dashboards can be operational-focused, functionality/business-focused, or a combination of both. The monitoring and observability (telemetry) data should enable reports on two main aspects of a data system/process:

Reviewer: ```suggestion
Customers heavily rely on dashboards and metrics to evaluate the **business value** offered by a service or process. Therefore, it is crucial to capture all the necessary information at the appropriate granularity and frequency in order to generate these customer-facing dashboards. These dashboards can be operational-focused, functionality/business-focused, or a combination of both.   

The monitoring and observability (telemetry) data should enable reports on two main aspects of a data system/process:
```

1. *One time development activity* - Create [OTEL collector code](#sample-python-script---using-opentelemetry-collector) packaged as a common library which creates OpenTelemetry providers for traces, metrics and logs using the Collector.

1. *Regular development activity* - Use common OTEL library, from above step, in the code to [capture observability data](https://opentelemetry.io/docs/languages/python/instrumentation/).

   - Understand different ways to [create spans](#create-tracesspans).
   - The [code sample](#creating-telemetry-events-with-opentelemetry-collector) shows a way to generate the telemetry.

### Create telemetry data

The following explains different types of OTEL spans and ways to create them and the [samples](#sample-files) contains examples of a [common schema](#sample-schema) and combining business data with telemetry data.

> #### Understand OpenTelemetry span types

Reviewer: ```suggestion
#### Understand Span Types
```

> Spans are one type of [signals](https://opentelemetry.io/docs/concepts/signals/) generated by OpenTelemetry. There are primarily [two types of Spans](https://opentelemetry-python.readthedocs.io/en/latest/api/trace.html):

Reviewer: ```suggestion
There are primarily [two types of Spans](https://opentelemetry-python.readthedocs.io/en/latest/api/trace.html):
```

    LoggingHandler,
)
from opentelemetry.sdk._logs.export import BatchLogRecordProcessor
from opentelemetry.sdk.resources import SERVICE_NAME, Resource
# metrics
from opentelemetry import metrics
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader

logging.basicConfig(level=logging.INFO)

class OpenTelemetryAppInsightsExporter:
    def __init__(self, conn_string: str) -> None:
        """
        Initializes the OpenTelemetryExporter class.

        Args:
            conn_string (str): Azure AppInsights connection string.
        """
        self.conn_string = conn_string

        return None

    def get_otel_tracer(self, trace_resource_attributes: dict, tracer_name: str = __name__) -> object:
        """
        Creates and returns an OpenTelemetry tracer object.

        Args:
            trace_resource_attributes (dict): The OpenTelemetry resource attributes in dictionary format
            tracer_name (str): The name of the tracer. Default is __name__.
        Returns:
            tracer: OpenTelemetry tracer object
        """
        resource = Resource(attributes=trace_resource_attributes)

>         tracer_provider = TracerProvider(resource=resource)
>         # Exporter to send data to AppInsights
>         trace_exporter = AzureMonitorTraceExporter(connection_string=self.conn_string)
>         span_processor = BatchSpanProcessor(trace_exporter)
>         tracer_provider.add_span_processor(span_processor)
>         tracer = trace.get_tracer(tracer_name, tracer_provider=tracer_provider)
> 
>         return tracer
> 
> 
>     def get_otel_logger(self, log_resource_attributes: dict, logger_name: str = __name__, add_console_handler: bool = True) -> object:
>         """
>         Creates and returns an OpenTelemetry logger object.
> 
>         Args:
>             log_resource_attributes (dict): The OpenTelemetry resource attributes in dictionary format
>             logger_name (str): The name of the logger. Default is __name__.
>             add_console_handler (bool): Whether to add a console handler to the logger. Default is True.
>         Returns:
>             logger: OpenTelemetry logger object
>         """
>         resource = Resource(attributes=log_resource_attributes)
>         log_exporter = AzureMonitorLogExporter(connection_string=self.conn_string)
> 
>         logger_provider = LoggerProvider(resource=resource)
>         logger_provider.add_log_record_processor(BatchLogRecordProcessor(log_exporter))
>         handler = LoggingHandler(level=logging.INFO, logger_provider=logger_provider)
>         logging.getLogger().addHandler(handler) # Attach OTLP handler to root logger
> 
>         logger = logging.getLogger(logger_name) # get namespaced logger
> 
>         # # Create a console handler  - Optional
>         if add_console_handler:
>             console_handler = logging.StreamHandler()
>             console_handler.setLevel(logging.INFO)
>             logger.addHandler(console_handler)
> 
>         return logger
> 
> 
>     def get_otel_metrics(self, metric_resource_attributes: dict, metric_name: str = __name__, metric_version: str = "0") -> object:
>         """
>         Creates and returns an OpenTelemetry metrics object.
> 
>         Args:
>             metric_resource_attributes (dict): The OpenTelemetry resource attributes in dictionary format
>             metric_name (str): The name of the metric. Default is __name__.
>             metric_version (str): The version of the metric. Default is "0".
>         Returns:
>             meter: OpenTelemetry meter object
>         """
>         resource = Resource(attributes=metric_resource_attributes)
>         metrics_exporter = AzureMonitorMetricExporter(connection_string=self.conn_string)
>         metrics_reader = PeriodicExportingMetricReader(metrics_exporter)
>         metrics_provider = MeterProvider(resource=resource, metric_readers=[metrics_reader])
>         metrics.set_meter_provider(metrics_provider)
>         meter = metrics.get_meter_provider().get_meter(name=metric_name, version=metric_version)
> 
>         return meter
> ```

Reviewer 2: ```suggestion
[Sample code (otel_monitor_invoker.py)](https://github.com/Azure-Samples/modern-data-warehouse-dataops/blob/feat/e2e-fabric-dataops-sample/e2e_samples/fabric_dataops_sample/config/fabric_environment/otel_monitor_invoker.py)
```

#### Creating telemetry events with OpenTelemetry SDK (No Collector)

See [Creating traces and spans](#create-tracesspans) for typical flow.

Assuming [the code for SDK option](#sample-python-script-with-opentelemetry-sdk-no-collector-using-azure-monitor-as-the-sink) is stored `otel_monitor_invoker.py` here is the sample code to use for generating telemetry.

```python

from opentelemetry.trace import SpanKind
from opentelemetry.trace.status import StatusCode

import otel_monitor_invoker as otel  # OTEL code saved as the code module

otlp_exporter = otel.OpenTelemetryAppInsightsExporter(conn_string=<< replace this with Azure Monitor connection string>>)

# set resources - This is only a sample. Make it detailed enough so that SRE/OPS have information for their monitoring needs.
resource_attributes = {
    # ---------- Reserved attribute names
    "service.name": "your service_name",
    "service.version": "your service_version",

>     "service.namespace": "your name space",
>     "service.instance.id": " your job execution id",
>     "process.executable.name": "business friendly process name",
>     "deployment.environment": "environment name",
>     # ---------- custom attributes - we can also add common attributes like appid, domain id etc
>     # Here "jobexec.*" is a custom namespace to denote all job execution information.
>     "jobexec.context": "your execution context"
>     "jobexec.cluster.region": "your region name",
>     "jobexec.app.name": "your app name",
>     "jobexec.instance.name": "your unique job execution instance id",
> }
> 
> # Typically, logging is performed within the context of a span.
> #   This allows log messages to be associated with trace information through the use of trace IDs and span IDs.
> #   As a result, it's generally not necessary to include resource information in log messages.
> # Note that trace IDs and span IDs will be null when logging is performed outside of a span context.
> log_attributes = {"jobexec.instance.name": "your unique job execution instance id"}
> trace_attributes = resource_attributes
> 
> # Only showing tracer and logger but not metrics. But the concept is same.
> tracer = otlp_exporter.get_otel_tracer(trace_resource_attributes=trace_attributes)
> logger = otlp_exporter.get_otel_logger(
>     log_resource_attributes=log_attributes,
>     logger_name="parking-logger",
>     add_console_handler=False,
> )
> logger.setLevel("INFO")
> 
> # Use tracer and logger as needed
> 
> root_span_name = <<create a unique rootname - make it user friendly for easy querying>>
> 
> with tracer.start_as_current_span(root_span_name, kind=SpanKind.INTERNAL) as root_span:
>     try:
>         # sample event addition
>         root_span.add_event(
>             name="010_first_Step", # some name for this particular event
>             attributes={"key1": value1, "Key2": value2}, # custom attributes related to first_Step as elements of a dictionary
>         )
>         << code related to first_step>>
> 
>         << create child spans as needed >>
> 
>         << add as many events and spans as required>>
> 
>     except Exception as e:
>         error_message = "<<your error message >> Error is:\n" + e
>         logger.exception(error_message)
>         root_span.set_status(StatusCode.ERROR, error_message) # set failure status on root level in case of failures
>         root_span.record_exception(e)
>         raise # raise or continue as needed
>     else:
>         root_span.set_status(StatusCode.OK) # set success status on root level in case of success
>         logger.info("Setup process is successful.")
>     finally:
>         logger.info(
>             "\n** Setup process is complete. Check the logs for execution status. End of setup process. **\n\n"
>         )
> 
> ```

Reviewer 2: ```suggestion
[Notebook Sample (00_setup.ipynb)](https://github.com/Azure-Samples/modern-data-warehouse-dataops/blob/feat/e2e-fabric-dataops-sample/e2e_samples/fabric_dataops_sample/src/notebooks/00_setup.ipynb)
```

        Args:
            metric_resource_attributes (dict): The OpenTelemetry resource attributes in dictionary format
            metric_name (str): The name of the metric. Default is __name__.
            metric_version (str): The version of the metric. Default is "0".
        Returns:
            meter: OpenTelemetry meter object
        """
        resource = Resource(attributes=metric_resource_attributes)
        metrics_exporter = OTLPMetricExporter(
            endpoint=self.metrics_end_point
            )
        metrics_reader = PeriodicExportingMetricReader(metrics_exporter)
        metrics_provider = MeterProvider(resource=resource, metric_readers=[metrics_reader])
        metrics.set_meter_provider(metrics_provider)
        meter = metrics.get_meter_provider().get_meter(name=metric_name, version=metric_version)

        return meter
```

#### Creating telemetry events with OpenTelemetry Collector

See [Creating traces and spans](#create-tracesspans) for typical flow.

Assuming [the code for Collector option](#sample-python-script---using-opentelemetry-collector) is stored as `otel_collector_invoker.py` here is the sample code to use for generating telemetry.

> - The code will be same as [telemetry events generation using SDK](#creating-telemetry-events-with-opentelemetry-sdk-no-collector) after replacing the the two lines as shown below:

Reviewer 2: ```suggestion
- The code will be same as [telemetry events generation using SDK](https://github.com/Azure-Samples/modern-data-warehouse-dataops/blob/feat/e2e-fabric-dataops-sample/e2e_samples/fabric_dataops_sample/src/notebooks/00_setup.ipynb) after replacing the two lines as shown below:
```

**Where data is sent**:

Given the sample configuration file contents below:

- For [ADX](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/azuredataexplorerexporter) data will be seen in `oteldb` (See [ADX DDLs](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/azuredataexplorerexporter#database-and-table-definition-scripts) ). Table names will be `OTELTraces` for Traces, `OTELLogs` for logs and `OTELMetrics` for metrics.
- For [App Insights/Azure monitor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/azuremonitorexporter)
  - Data will be seen in as follows:
    - Traces:
      - "dependency" for Spankind in [CLIENT, PRODUCER, INTERNAL]
      - "requests" for Spankind in [SERVER, CONSUMER]
    - Logs: "traces"
    - Metrics: "customMetrics"
    - Exception events: "exceptions"

The sample Collector configuration below uses an Environment file(`env.cfg`) placed on the machine where Collector is setup.

```bash
####################
# file_name: env.cfg
#####################
# Azure App Insights

> APPINSIGHTS CONNECTION STRING="<<App-insgights-connection-string>>"

Reviewer 2: Locally tested and double quote (`"`) will be treated as a part of string, so need to remove
```suggestion
APPINSIGHTS CONNECTION STRING=<<App-insgights-connection-string>>
```
